### ISSUE TITLE
Segfault during S3 upload

### ISSUE THREAD

**Issue by mariadb-nickvenenga** [fluent/fluent-bit#4369]
## Bug Report

**Describe the bug**
We encountered a segfault after fluent-bit had been running a while

**To Reproduce**
It's only happened once so far during our initially testing--I'm working on additional testing and will update if I can find a way to reproduce--I expect GCS responded with a large HTML 500 error page

**Expected behavior**
fluent-bit handles network issues without crashing

**Your Environment**

* Version used: master@10c3cd58658c20b989db859758a249109041070c
Built using this Docker image https://github.com/fluent/fluent-bit-docker-image/blob/master/Dockerfile.x86_64-master with FLB_TRACE=On
* Configuration:
```
  custom_parsers.conf: |
    [PARSER]
        Name docker_no_time
        Format json
        Time_Keep Off
        Time_Key time
        Time_Format %Y-%m-%dT%H:%M:%S.%L

  fluent-bit.conf: |
    [SERVICE]
        Flush 1
        Daemon Off
        Log_Level trace
        Parsers_File parsers.conf
        Parsers_File custom_parsers.conf
        HTTP_Server On
        HTTP_Listen 0.0.0.0
        HTTP_Port 2020

    [INPUT]
        Name tail
        Path /var/log/containers/*.log
        Parser docker
        Tag kube.*
        Mem_Buf_Limit 5MB
        Skip_Long_Lines On

    [INPUT]
        Name systemd
        Tag host.*
        Systemd_Filter _SYSTEMD_UNIT=kubelet.service
        Read_From_Tail On

    [INPUT]
        Name tail
        Path /var/log/containers/nickv-log-test*.log
        Parser docker
        Tag mariadb.<namespace_name>.<pod_name>.<container_name>
        Tag_Regex (?<pod_name>[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*)_(?<namespace_name>[^_]+)_(?<container_name>.+)-
        Mem_Buf_Limit 5MB
        DB /var/log/containers/nickv-log-test.db
        DB.locking true
        DB.journal_mode WAL
        Offset_Key offset

    [FILTER]
        Name kubernetes
        Match kube.*
        Merge_Log On
        Keep_Log Off
        K8S-Logging.Parser On
        K8S-Logging.Exclude On

    [FILTER]
        Name lua
        Match kube.*
        script /fluent-bit/etc/skysql_scripts.lua
        call dedot

    [OUTPUT]
        name s3
        match mariadb.*
        bucket <redacted>
        region us-central1
        endpoint https://storage.googleapis.com
        total_file_size 5M
        upload_timeout 5m
        use_put_object On
        s3_key_format /sky0005922/$TAG[1]/nickv-log-test/$TAG[2]/%Y/%Y_%m_%d/error-log_%Y-%m-%d_$UUID.log
        s3_key_format_tag_delimiters .
```
(I had net.keepalive=off, net.tcp_keepalive=off options incorrectly specified in [SERVICE] I don't still have a copy of in addition to this config)

* Environment name and version (e.g. Kubernetes? What version?): GKE v1.20.10-gke.1600
* Server type and version: COS docker image
* Operating System and version:
* Filters and plugins:
S3 (See config)

**Additional context**
```
[2021/11/24 01:49:10] [engine] caught signal (SIGSEGV)
#0  0x563bc4736f61      in  flb_io_net_read() at src/flb_io.c:412
#1  0x563bc4739577      in  flb_http_do() at src/flb_http_client.c:1206
#2  0x563bc4814cc2      in  request_do() at src/aws/flb_aws_util.c:413
#3  0x563bc4814198      in  flb_aws_client_request() at src/aws/flb_aws_util.c:160
#4  0x563bc47dff96      in  s3_put_object() at plugins/out_s3/s3.c:1386
#5  0x563bc47decf3      in  upload_data() at plugins/out_s3/s3.c:1028
#6  0x563bc47e1654      in  cb_s3_upload() at plugins/out_s3/s3.c:1783
#7  0x563bc47179ba      in  flb_sched_event_handler() at src/flb_scheduler.c:422
#8  0x563bc4714550      in  flb_engine_start() at src/flb_engine.c:765
#9  0x563bc46f683f      in  flb_lib_worker() at src/flb_lib.c:627
#10 0x7fba6c522fa2      in  ???() at ???:0
#11 0x7fba6bc034ce      in  ???() at ???:0
#12 0xffffffffffffffff  in  ???() at ???:0
```

Here's a log with some additional context--it looks like it's receiving a lot of data before the crash
[fb-1.9.0-crash.log](https://github.com/fluent/fluent-bit/files/7597322/fb-1.9.0-crash.log)

**Comment by mariadb-nickvenenga** [comment]
Looks like it happened again after ~4 hours
```
[2021/11/24 18:55:57] [engine] caught signal (SIGSEGV)
[2021/11/24 18:55:57] [trace] [io coro=0x7f01fcadc060] [net_read] try up to 34 bytes
[2021/11/24 18:55:57] [trace] [io coro=0x7f01fcadc060] [net_read] ret=5
[2021/11/24 18:55:57] [trace] [io coro=0x7f01fcadc060] [net_read] try up to 29 bytes
[2021/11/24 18:55:57] [trace] [io coro=0x7f01fcadc060] [net_read] ret=5
[2021/11/24 18:55:57] [trace] [io coro=0x7f01fcadc060] [net_read] try up to 24 bytes
[2021/11/24 18:55:57] [trace] [io coro=0x7f01fcadc060] [net_read] ret=5
[2021/11/24 18:55:57] [trace] [io coro=0x7f01fcadc060] [net_read] try up to 19 bytes
[2021/11/24 18:55:57] [trace] [io coro=0x7f01fcadc060] [net_read] ret=5
[2021/11/24 18:55:57] [trace] [io coro=0x7f01fcadc060] [net_read] try up to 14 bytes
[2021/11/24 18:55:57] [trace] [io coro=0x7f01fcadc060] [net_read] ret=5
[2021/11/24 18:55:57] [trace] [io coro=0x7f01fcadc060] [net_read] try up to 9 bytes
[2021/11/24 18:55:57] [trace] [io coro=0x7f01fcadc060] [net_read] ret=5
[2021/11/24 18:55:57] [trace] [io coro=0x7f01fcadc060] [net_read] try up to 4 bytes
[2021/11/24 18:55:57] [trace] [io coro=0x7f01fcadc060] [net_read] ret=5
[2021/11/24 18:55:57] [trace] [io coro=0x7f01fcadc060] [net_read] try up to 32767 bytes
#0  0x556ec6734f61      in  flb_io_net_read() at src/flb_io.c:412
#1  0x556ec6737577      in  flb_http_do() at src/flb_http_client.c:1206
#2  0x556ec6812cc2      in  request_do() at src/aws/flb_aws_util.c:413
#3  0x556ec6812198      in  flb_aws_client_request() at src/aws/flb_aws_util.c:160
#4  0x556ec67ddf96      in  s3_put_object() at plugins/out_s3/s3.c:1386
#5  0x556ec67dccf3      in  upload_data() at plugins/out_s3/s3.c:1028
#6  0x556ec67dee70      in  send_upload_request() at plugins/out_s3/s3.c:1631
#7  0x556ec67e0a75      in  cb_s3_flush() at plugins/out_s3/s3.c:2171
#8  0x556ec67010c2      in  output_pre_cb_flush() at include/fluent-bit/flb_output.h:511
#9  0x556ec6bfe226      in  co_init() at lib/monkey/deps/flb_libco/amd64.c:117
```

**Comment by nokute78** [comment]
```
[2021/11/24 18:55:57] [trace] [io coro=0x7f01fcadc060] [net_read] try up to 4 bytes
[2021/11/24 18:55:57] [trace] [io coro=0x7f01fcadc060] [net_read] ret=5
```
It is strange.
https://github.com/fluent/fluent-bit/blob/v1.8.10/src/flb_io.c#L387-L408
The read size is 5 in spite of the buffer size is 4.

I don't know why it was happened...

**Comment by mariadb-nickvenenga** [comment]
Also seeing a slightly different stacktrace running in gdb (I don't know much about gdb)

```
[2021/12/06 22:29:48] [trace] [io coro=0x7ffff55ef630] [net_read] try up to 26291 bytes
[2021/12/06 22:29:48] [trace] [io coro=0x7ffff55ef630] [net_read] ret=5
[2021/12/06 22:29:48] [trace] [io coro=0x7ffff55ef630] [net_read] try up to 26286 bytes
[2021/12/06 22:29:48] [trace] [io coro=0x7ffff55ef630] [net_read] ret=5
[2021/12/06 22:29:48] [trace] [upstream] destroy connection #80 to storage.googleapis.com:443
[2021/12/06 22:29:48] [debug] [output:s3:s3.0] PutObject http status=200
[2021/12/06 22:29:48] [ info] [output:s3:s3.0] Successfully uploaded object /obj.log
[2021/12/06 22:29:48] [debug] [fstore] [cio file] synced at: 2021-12-06T20:43:53/704190038316138288-17188832368491895580
[2021/12/06 22:29:48] [trace] [engine] [task event] task_id=3 out_id=0 return=OK
[2021/12/06 22:29:48] [debug] [out flush] cb_destroy coro_id=1255
[2021/12/06 22:29:48] [trace] [coro] destroy coroutine=0x7ffff55ef6d8 data=0x7ffff55ef6f0
[2021/12/06 22:29:48] [debug] [out flush] cb_destroy coro_id=1501
[2021/12/06 22:29:48] [trace] [coro] destroy coroutine=0x7ffff55867a8 data=0x7ffff55867c0
[2021/12/06 22:29:48] [debug] [out flush] cb_destroy coro_id=1258
[2021/12/06 22:29:48] [trace] [coro] destroy coroutine=0x7ffff5586880 data=0x7ffff5586898
[2021/12/06 22:29:48] [debug] [out flush] cb_destroy coro_id=1262
[2021/12/06 22:29:48] [trace] [coro] destroy coroutine=0x7ffff55ef630 data=0x7ffff55ef648
[2021/12/06 22:29:48] [debug] [task] destroy task=0x7ffff54463b0 (task_id=3)
[2021/12/06 22:29:48] [debug] [socket] could not validate socket status for #77 (don't worry)
[2021/12/06 22:29:48] [debug] [socket] could not validate socket status for #80 (don't worry)

Thread 2 "flb-pipeline" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffff61ff700 (LWP 49)]
je_large_dalloc (tsdn=tsdn@entry=0x7ffff61f54a0, extent=0x0) at /source/lib/jemalloc-5.2.1/src/large.c:361
361             arena_t *arena = extent_arena_get(extent);
(gdb) t a a bt

Thread 6 (Thread 0x7fffe85fe700 (LWP 53)):
#0  0x00007ffff76917ef in epoll_wait () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x0000555555b65f1d in _mk_event_wait (loop=0x7fffe7815000) at /source/lib/monkey/mk_core/mk_event_epoll.c:366
#2  0x0000555555b66232 in mk_event_wait (loop=0x7fffe7815000) at /source/lib/monkey/mk_core/mk_event.c:169
#3  0x0000555555b5f979 in mk_server_worker_loop (server=0x7ffff54c0b00) at /source/lib/monkey/mk_server/mk_server.c:470
#4  0x0000555555b56ceb in mk_sched_launch_worker_loop (data=0x7fffe8e00008) at /source/lib/monkey/mk_server/mk_scheduler.c:416
#5  0x00007ffff7fb0fa3 in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#6  0x00007ffff76914cf in clone () from /lib/x86_64-linux-gnu/libc.so.6

Thread 5 (Thread 0x7fffe8dff700 (LWP 52)):
#0  0x00007ffff765e720 in nanosleep () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x00007ffff765e62a in sleep () from /lib/x86_64-linux-gnu/libc.so.6
#2  0x0000555555b5e8c6 in mk_clock_worker_init (data=0x7ffff54c0b00) at /source/lib/monkey/mk_server/mk_clock.c:132
#3  0x00007ffff7fb0fa3 in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#4  0x00007ffff76914cf in clone () from /lib/x86_64-linux-gnu/libc.so.6

Thread 4 (Thread 0x7fffe9bfe700 (LWP 51)):
#0  0x00007ffff76917ef in epoll_wait () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x0000555555b65f1d in _mk_event_wait (loop=0x7ffff5586d78) at /source/lib/monkey/mk_core/mk_event_epoll.c:366
#2  0x0000555555b66232 in mk_event_wait (loop=0x7ffff5586d78) at /source/lib/monkey/mk_core/mk_event.c:169
#3  0x0000555555b4eb31 in mk_lib_worker (data=0x7ffff5586d90) at /source/lib/monkey/mk_server/mk_lib.c:154
#4  0x00007ffff7fb0fa3 in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#5  0x00007ffff76914cf in clone () from /lib/x86_64-linux-gnu/libc.so.6

Thread 3 (Thread 0x7ffff53ff700 (LWP 50)):
#0  0x00007ffff76917ef in epoll_wait () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x0000555555b65f1d in _mk_event_wait (loop=0x7ffff540a018) at /source/lib/monkey/mk_core/mk_event_epoll.c:366
#2  0x0000555555b66232 in mk_event_wait (loop=0x7ffff540a018) at /source/lib/monkey/mk_core/mk_event.c:169
#3  0x000055555565efb6 in log_worker_collector (data=0x7ffff542b000) at /source/src/flb_log.c:132
#4  0x0000555555684898 in step_callback (data=0x7ffff54370a0) at /source/src/flb_worker.c:44
#5  0x00007ffff7fb0fa3 in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#6  0x00007ffff76914cf in clone () from /lib/x86_64-linux-gnu/libc.so.6

Thread 2 (Thread 0x7ffff61ff700 (LWP 49)):
#0  je_large_dalloc (tsdn=tsdn@entry=0x7ffff61f54a0, extent=0x0) at /source/lib/jemalloc-5.2.1/src/large.c:361
#1  0x00005555555ed02d in arena_dalloc_large (slow_path=<optimized out>, szind=<optimized out>, tcache=<optimized out>, ptr=<optimized out>, tsdn=<optimized out>)
    at /source/lib/jemalloc-5.2.1/include/jemalloc/internal/arena_inlines_b.h:281
#2  arena_dalloc (slow_path=<optimized out>, alloc_ctx=<optimized out>, tcache=<optimized out>, ptr=<optimized out>, tsdn=<optimized out>)
    at /source/lib/jemalloc-5.2.1/include/jemalloc/internal/arena_inlines_b.h:323
#3  idalloctm (slow_path=<optimized out>, is_internal=<optimized out>, alloc_ctx=<optimized out>, tcache=<optimized out>, ptr=<optimized out>, tsdn=<optimized out>)
    at /source/lib/jemalloc-5.2.1/include/jemalloc/internal/jemalloc_internal_inlines_c.h:118
#4  ifree (slow_path=<optimized out>, tcache=<optimized out>, ptr=<optimized out>, tsd=<optimized out>) at /source/lib/jemalloc-5.2.1/src/jemalloc.c:2589
#5  je_free_default (ptr=<optimized out>) at /source/lib/jemalloc-5.2.1/src/jemalloc.c:2799
#6  0x00007ffff7c35a8f in ?? () from /usr/lib/x86_64-linux-gnu/libcrypto.so.1.1
#7  0x00007ffff7c35ef0 in ?? () from /usr/lib/x86_64-linux-gnu/libcrypto.so.1.1
#8  0x00007ffff7c35bb6 in ?? () from /usr/lib/x86_64-linux-gnu/libcrypto.so.1.1
#9  0x00007ffff7c35ef0 in ?? () from /usr/lib/x86_64-linux-gnu/libcrypto.so.1.1
#10 0x00007ffff7c35bb6 in ?? () from /usr/lib/x86_64-linux-gnu/libcrypto.so.1.1
#11 0x00007ffff7c35e15 in ASN1_item_free () from /usr/lib/x86_64-linux-gnu/libcrypto.so.1.1
#12 0x00007ffff7d73390 in OPENSSL_sk_pop_free () from /usr/lib/x86_64-linux-gnu/libcrypto.so.1.1
#13 0x00007ffff7ea7710 in SSL_free () from /usr/lib/x86_64-linux-gnu/libssl.so.1.1
#14 0x00005555556934d4 in tls_session_destroy (session=0x7ffff55ef678) at /source/src/tls/openssl.c:338
#15 0x0000555555694109 in flb_tls_session_destroy (tls=0x7ffff5583a80, u_conn=0x7ffff542c980) at /source/src/tls/flb_tls.c:427
#16 0x0000555555682b7a in destroy_conn (u_conn=0x7ffff542c980) at /source/src/flb_upstream.c:460
#17 0x0000555555683a8e in flb_upstream_conn_pending_destroy (u=0x7ffff54d4600) at /source/src/flb_upstream.c:890
#18 0x0000555555683be5 in flb_upstream_conn_pending_destroy_list (list=0x7ffff66332e0) at /source/src/flb_upstream.c:940
#19 0x000055555567c920 in flb_engine_start (config=0x7ffff6632fc0) at /source/src/flb_engine.c:800
#20 0x000055555565e840 in flb_lib_worker (data=0x7ffff660b340) at /source/src/flb_lib.c:627
#21 0x00007ffff7fb0fa3 in start_thread () from /lib/x86_64-linux-gnu/libpthread.so.0
#22 0x00007ffff76914cf in clone () from /lib/x86_64-linux-gnu/libc.so.6

Thread 1 (Thread 0x7ffff6a75c40 (LWP 45)):
#0  0x00007ffff765e720 in nanosleep () from /lib/x86_64-linux-gnu/libc.so.6
#1  0x00007ffff765e62a in sleep () from /lib/x86_64-linux-gnu/libc.so.6
#2  0x00005555555e41d9 in flb_main (argc=4, argv=0x7fffffffe3f8) at /source/src/fluent-bit.c:1286
#3  0x00005555555e4250 in main (argc=4, argv=0x7fffffffe3f8) at /source/src/fluent-bit.c:1303
```

Seems potentially related to #3331

**Comment by shaojielinux** [comment]
I also encountered the same problem. Have you solved it?

**Comment by nokute78** [comment]
Similar issue is reported https://github.com/aws/aws-for-fluent-bit/issues/278#issuecomment-994968858
The issue points https://github.com/fluent/fluent-bit/issues/4098

**Comment by nokute78** [comment]
The patch #4100 for https://github.com/fluent/fluent-bit/issues/4369#issuecomment-978157281 is merged into master.
The patch #4584 is for v1.8.x branch and it is not merged yet.

**Comment by nokute78** [comment]
The patch #4584 is merged from v1.8.12.
